uxsdcxx/capnp.py

```diff
@@ -301,6 +301,190 @@ def load_fn_from_complex_type(t: UxsdComplex) -> str:
+				input="in.get_{name}(i, {data}, {itr})".format(
+					name=cpp._gen_stub_suffix(t, parent),
+					data=data,
+					itr=itr))
+		else:
+			return _gen_set_simple(
+				t.type,
+				input="in.get_{name}({data}, {itr})".format(
+					name=cpp._gen_stub_suffix(t, parent),
+					data=data,
+					itr=itr))
+	else:
+		raise NotImplementedError(t)
+
+
+def _gen_write_complex_element(e: UxsdElement, parent: str) -> str:
+	"""Function to generate partial code which writes out an element with a complex type."""
+	assert isinstance(e.type, UxsdComplex)
+	out = ""
+
+	def _gen_write_element_body(root : str) -> str:
+		assert isinstance(e.type, UxsdComplex)
+		ouv = ""
+		if e.type.attrs:
+			for a in e.type.attrs:
+				ouv += _gen_write_attr(a, e.type.name, root, "child_data", "child_iter")
+			if e.type.content:
+				ouv += "write_{name}_capnp_type(in, {root}, child_data, child_iter);\n".format(
+						root=root,
+						name=e.type.name)
+		else:
+			if e.type.content:
+				ouv += "write_{name}_capnp_type(in, {root}, child_data, child_iter);\n".format(
+						root=root,
+						name=e.type.name)
+
+		return ouv
+
+	name = cpp._gen_stub_suffix(e, parent)
+	if e.many:
+		plural_name = utils.pluralize(cpp._gen_stub_suffix(e, parent))
+		out += "size_t num_{pname} = in.num_{name}(data, iter);\n".format(name=name, pname=plural_name)
+		out += "auto {name} = root.init{pname}(num_{name});\n".format(
+				name=plural_name,
+				pname=utils.pluralize(utils.to_pascalcase(e.name)))
+		out += "for(size_t i = 0; i < num_{name}; i++) {{\n".format(name=plural_name)
+		out += "\tauto {name} = {plural_name}[i];\n".format(
+				name=name,
+				plural_name=plural_name,
+				)
+		out += "\tstd::tie(child_data, child_iter) = in.get_%s(i, data, iter);\n" % cpp._gen_stub_suffix(e, parent)
+		out += utils.indent(_gen_write_element_body(name))
+		out += "}\n"
+	elif e.optional:
+		out += "if(in.has_%s(data, iter)){\n" % cpp._gen_stub_suffix(e, parent)
+		out += "\tauto {name} = root.init{pname}();\n".format(
+				name=name,
+				pname=utils.to_pascalcase(e.name))
+		out += "\tstd::tie(child_data, child_iter) = in.get_%s(data, iter);\n" % cpp._gen_stub_suffix(e, parent)
+		out += utils.indent(_gen_write_element_body(name))
+		out += "}\n"
+	else:
+		out += "std::tie(child_data, child_iter) = in.get_%s(data, iter);\n" % cpp._gen_stub_suffix(e, parent)
+		out += "auto {name} = root.init{pname}();\n".format(
+				name=name,
+				pname=utils.to_pascalcase(e.name))
+		out += _gen_write_element_body(name)
+
+	return out
+
+
+def _gen_write_element(e: UxsdElement, parent: str) -> str:
+	"""Function to generate partial C++ code for writing out a struct generated
+	from an UxsdElement.
+
+	Currently, all values with non-zero default values are emitted.
+	Otherwise, we would have to check against the nonzero value, and the
+	check would create a case split for all simple types again.(how to compare
+	unions? strings? doubles?)
+	"""
+	out = ""
+	if isinstance(e.type, UxsdSimple):
```

Comment:
I think this was left over from the XML fn?

Reply:
Fixed